### PR TITLE
Introduce `AdwButton` widget

### PR DESCRIPTION
**Pull request by jesusrp98**
This PR introduces a new widget `RawAdwTarget`. This widget could be useful when you want to build a button or any clickable widget: `AdwSidebarItem`, `AdwTextButton` etc.

- It handles its state internally (hover, enabled, active) and allows you to change its background color and child depending on the state.
- This is based on the work done on the `AdwHeaderButton` and various other widgets around the app.
- It allows to reuse common logic and system for this kind of situations.

Included in this PR is a rework on the `AdwHeaderButton` using this new widget, in order to illustrate how this new widget could be used.

Let me know what you think about this widget and its use in the app.

### CODE REVIEW

File: lib/src/widgets/adw/raw_target.dart
Comment (jesusrp98) [inline lib/src/widgets/adw/raw_target.dart:9]: I'll add some documentation if you want to proceed with this change :)

**Comment by jesusrp98** [comment]
My plan is to build all the available buttons inside GTK's libadwaita.

image

**Comment by prateekmedia** [comment]
Looks good!

File: lib/src/widgets/adw/new/header_button.dart
Comment (MalcolmMielle) [inline lib/src/widgets/adw/new/header_button.dart:6]: Should the parameter be called child?

File: lib/src/widgets/adw/new/header_button.dart
Comment (MalcolmMielle) [inline lib/src/widgets/adw/new/header_button.dart:9]: I would rename `icon` to `child` since it can be any widget, although I know it was named `icon` before so it's not your fault!

Comment (jesusrp98) [reply lib/src/widgets/adw/new/header_button.dart:9]: I'd like to rename the `icon` parameter to `child`, makes sense.

I didn't want to introduce breaking changes in this PR, but I can create a future PR to resolve this.

**Comment by jesusrp98** [comment]
Just added a couple more parameters to the `RawAdwTarget`: `boxShadows`, `shape` and `border`.

They could be useful in the future when using this class :)

File: lib/src/widgets/adw/new/header_button.dart
Comment (MalcolmMielle) [reply lib/src/widgets/adw/new/header_button.dart:9]: That sounds great.

File: lib/src/widgets/adw/raw_target.dart
Comment (MalcolmMielle) [inline lib/src/widgets/adw/raw_target.dart:16]: Do you see this class being used for more than just buttons? Could it/would it be used for stuck like "AdwComboBox"? if no I would rename the class to something more "understandable" such as `AdwBaseButton`.

I would also propose to clarify the documentation a little bit. The code is very clear, easy to understand, and well made but the documentation sounded vague to me. What about something like:

```dart
/// RawAdwTarget is a widget used as a base to build Adwaita widgets
/// with pressed and hover properties.
///
/// RawAdwTarget widgets can rebuild their background through
/// the [this.backgroundColorBuilder] callback.
```

File: lib/src/widgets/adw/raw_target.dart
Comment (MalcolmMielle) [inline lib/src/widgets/adw/raw_target.dart:21]: Do we want this to be `@required`? What if the user simply wants the default background but still want to use the gesture mechanism given by `_status`? Does that seems like an unlikely use of the class?

File: lib/src/widgets/adw/raw_target.dart
Comment (MalcolmMielle) [inline lib/src/widgets/adw/raw_target.dart:137]: I think that it may be a good idea to separate the hover/tap logic.

For example, let's say I use the class with a pop-up (the comment line 101 seems to indicate that it's a possibility), and I have `_status = TargetStatus.active`. If I hover the pop-up, I will loose my active status. That does not seem correct to me, since the widget could be active _and_ hover? Am I missing something :S?

From my opinion, we should have two state elements: _statushover and _statusActive. This way I can hover over an active pop-up. Does that sound right to you :)?

**Comment by MalcolmMielle** [comment]
Excellent commits. I have a couple of small remarks but overall it's a great contribution thanks :).

**Review by MalcolmMielle — CHANGES_REQUESTED** [review]

File: lib/src/widgets/adw/raw_target.dart
Comment (jesusrp98) [reply lib/src/widgets/adw/raw_target.dart:16]: I do think that this widget should be the base of all components that are 'clickable' by the user. This is useful to handle changes when a user hovers or clicks on the view.

So yes, This widget should support a widget like `AdwComboBox`. Maybe in the process of using this new widget inside the combo box we find some aspect that could be improved or changes.

I'll update the widget's documentation, since the one you've proposed looks better than the current one :)

File: lib/src/widgets/adw/raw_target.dart
Comment (jesusrp98) [reply lib/src/widgets/adw/raw_target.dart]: You're right here. It could be a use case where no color is needed, thus the `backgroundColorBuilder` parameter should be 'nullable'.

If so, no color will be painted in the background and it will be transparent.

File: lib/src/widgets/adw/raw_target.dart
Comment (jesusrp98) [reply lib/src/widgets/adw/raw_target.dart:137]: You're correct there. A `RawAdwTarget` derived widget cannot be active + hover at the same time, and it should be, because it's a real possibility.

I don't like the idea of having multiple variables to indicate a single thing like `_statusHover` or `_statusActive`. Having just a single variable, `_status` to denote the status of the widget is much more convenient when making the UI react according to the status.

The down side of using an enum here is that it cannot be two things at the same time. So I propose create a new status: `activeHover`, which will be activated when hovering an activated widget. This way, we can control that specific case, and not deal with multiple states that could be troublesome to handle, active, tap down and hovered statuses.

What do you think about this @MalcolmMielle ?

File: lib/src/widgets/adw/raw_target.dart
Comment (MalcolmMielle) [reply lib/src/widgets/adw/raw_target.dart:16]: Sounds good to me then :).
Let's keep in mind the migration of Combobox for later then!

File: lib/src/widgets/adw/raw_target.dart
Comment (MalcolmMielle) [reply lib/src/widgets/adw/raw_target.dart:137]: It might be a matter of semantic but I think `active` should be the isolated case. `hover`, `tapUp`, and `tapDown` are three states that cannot coexist and fit in `_status`. On the other hand, `isActive` is an independent state if I understand correclty.

Looking at the class I'm actually confused as to why `widget.active` make sense here? It isn't part of the `RawAdwTarget` state, and only make it so that `_status` has value `TargetStatus.active` or `TargetStatus.enabled`. If it is used in pop-up, I would say that it belongs as a self contained state, in another `RawWidget` type (e.g. `RawActivableAdwTarget` that would inherit from `RawAdwTarget`) where the _status can be: hover -> clickDown -> clickUp and a clickDown would turn _stateActive to true (so that the pop-up is fired). This RawActivableAdwTarget could have a `show` abstract method that triggers the pop-up (see AdwComboRow for an example of the `show` method).

This way we would have `RawAdwTarget` handling `hover`, `tapDown`, `TapUp`, and a new class `RawActivableAdwTarget` handling `isActive` and the trigger of the method associated with `active`.

I know it's more work, but I feel like this would be much cleaner and the integration with `ComboRow` would then be straight forward :D.

File: lib/src/widgets/adw/raw_target.dart
Comment (MalcolmMielle) [inline lib/src/widgets/adw/raw_target.dart:15]: I'm afraid the comment I'm about to make is a slight case of [the bike shed conundrum](https://bikeshed.com/), so feel free to disregard it and keep in mind that I think this commit is already very good! I won't be offended :).

I'm not familiar with any library naming interfaces/base widget `Raw`. This naming created a slight confusion for me when I read the commit, and it took me a couple of minutes to understand that it was an interface and not a type of widget I should use. I've tried to find an equivalent in libadwaita but couldn't find anything, and I think the closest is really [`GtkButton`](https://docs.gtk.org/gtk4/class.Button.html). I'm proposing to name such Widget `AdwButton`. Because in the end, that is what the class is, isn't? Other buttons that you plan to implement will all inherit from this class, right?

What do you think? If it's too much work for now, or you disagree, feel free to disregard this comment, but I think it would help with reading the code :).

Comment (jesusrp98) [reply lib/src/widgets/adw/raw_target.dart:15]: Thank you very much for the comment. I also think that `RawAdwTarget` could be a bit confusing.

After reading your comment, `AdwButton` looks like a good name to me as well. All 'clickable' elements are buttons in a way, so from that side makes sense.

I'll update the name shortly. And thanks for all the comments, this way we're improving the package immensely :)

Comment (MalcolmMielle) [reply lib/src/widgets/adw/raw_target.dart:15]: Sounds good to me :)

File: lib/src/widgets/adw/raw_target.dart
Comment (MalcolmMielle) [reply lib/src/widgets/adw/raw_target.dart:137]: Ok I think I understand your explaination. Let's aim for one `_status` variable then. We still need to be able to hover+active at the same time in the `_status`. I've just ran a quick test in libadwaita demo and the headerbar button darken on hover even when active. The funny thing is that ComboRow on the other hand, does **not** have a hover color when active, but it's ok with your implementation since the callback `_resolveBackgroundColor` is class independent :). A good point for your implementation ;).

I think that, then the simplest solution is to increase the range of possible states to incorporate active VS inactive states. We would then have `enum TargetStatus { enabledHover, activeHover, tapDown }` as possible states, and all possible state would be stored in `_status`. This change means that we would need to update the logic in `MouseRegion` and `GestureDetector`, but that logic would be fully integrated in the `AdwButton` so that's great for the user.

At that point we need to differentiate between activable and non-activable button to not clutter the logic of the Button Widget (since we will use it for pop-up and "normal" buttons).
I would either:

* Create two class `AdwButton` and `ActivableAdwbutton`, where `AdwButton` does not have any logic associated with activability.
* Add a optional parameter to `AdwButton` called `activable` that default to `false` when the button isn't activable.

I would prefer option 1, because `ActivableAdwbutton` will need a required `onActive` callback as you said, and having it as an optional parameter for `AdwButton` might be awkward (?). However, I don't have a strong opinion on that, and I'm also fine with option 2 if you prefer.

What do you think?

Comment (jesusrp98) [reply lib/src/widgets/adw/raw_target.dart:137]: Great! I'll change the `TargetStatus` variable to include the new `activeHovered` status and rename the other hover state. Thank you very much for the feedback :)

Lastly, I'd prefer to bundle all into a complete widget. If we have to different buttons we'd need to copy and past all the `MouseRegion` and `GestureDetector` logic, which won't be ideal, or create a even more basic widget with that logic and build the status on top of that.

I think having a `onActive` parameter inside `AdwButton` is much better. If `null`, then the button will not be 'activable'.

What do you think? We're so close!

Comment (MalcolmMielle) [reply lib/src/widgets/adw/raw_target.dart:137]: Sounds good to me! Let's go :D

Comment (jesusrp98) [reply lib/src/widgets/adw/raw_target.dart]: Just added the new states and updated the logic. It works super great and mimics completely the behavior that features the GTK's libadwaita.

In the end, the `isActive` and all the logic related to that feature should be handled outside the widget, here why.
- The 'active' state can be invoke when the user clicks on a button to show a pop up. In this case, baking all the 'active' state directly to the button widget would work great.
- However, there are more use cases where buttons are active. If you check the header bar of the example app, you can see a drawer 'hide' button that is activated, meaning that the drawer is visible.

This last behavior is not compatible with baking the active state directly to the button. We need an external dependency to tell the widget whether it's active or not.

This behavior is normal in the rest of material buttons and widgets.

Let me know what you think about this, but I think this is the only solution, having an interface to communicate the 'active' status of the button from outside. @MalcolmMielle

**Review by MalcolmMielle — APPROVED** [review]
Looks good to me! Congratulation and thanks for the PR.

**Comment by MalcolmMielle** [comment]
@prateekmedia could we merge #42 before merging this one so that the check will pass (unless you're almost done?). And then we'll do another PR to add the macos action. Does that sound good to you :)?

**Comment by jesusrp98** [comment]
> @prateekmedia could we merge #42 before merging this one so that the check will pass (unless you're almost done?). And then we'll do another PR to add the macos action. Does that sound good to you :)?

Sounds great! Please merge this PR once all pending ones are done :)

**Comment by MalcolmMielle** [comment]
@jesusrp98 I'm not sure what is going on with the linter at the moment :S. Could you try rebasing your branch on `main` to see if that helps?

**Comment by MalcolmMielle** [comment]
Nevermind I think the error is due to [this issue](https://github.com/github/super-linter/issues/507) ? it's surprising it passed before

**Comment by MalcolmMielle** [comment]
Should be solved by #44

**Comment by MalcolmMielle** [comment]
Bear with me @jesusrp98 can you rebase your commit, so the new linter is triggered :| ? If it does not work, I'll be so sad.

**Comment by MalcolmMielle** [comment]
Yes finally :D! Thanks a lot for everything @jesusrp98

**Pull request merged**
